packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+import {
+	useLayoutEffect,
+	useRef,
+	useMemo,
+	useEffect,
+	useReducer,
+} from '@wordpress/element';
+import { isShallowEqualObjects } from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+import useAsyncMode from '../async-mode-provider/use-async-mode';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the selector from the latest render. If a store update happens between render
+ * and the effect, this could cause missed/stale updates or inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+const useIsomorphicLayoutEffect =
+	typeof window !== 'undefined' ? useLayoutEffect : useEffect;
+
+const renderQueue = createQueue();
+
+export default function useSelect( mapSelect ) {
+	const registry = useRegistry();
+	const isAsync = useAsyncMode();
+	const queueContext = useMemo( () => ( { queue: true } ), [ registry ] );
+	const [ , forceRender ] = useReducer( ( s ) => s + 1, 0 );
+
+	const latestMapSelect = useRef();
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef();
+	const isMounted = useRef();
+
+	let mapOutput;
+
+	if ( latestMapSelect.current !== mapSelect ) {
+		mapOutput = mapSelect( registry.select, registry );
+	} else {
+		mapOutput = latestMapOutput.current;
+	}
+
+	useIsomorphicLayoutEffect( () => {
+		latestMapSelect.current = mapSelect;
+		if ( latestIsAsync.current !== isAsync ) {
+			latestIsAsync.current = isAsync;
+			renderQueue.flush( queueContext );
+		}
+		latestMapOutput.current = mapOutput;
+		isMounted.current = true;
+	} );
+
+	useIsomorphicLayoutEffect( () => {
+		const onStoreChange = () => {
+			if ( isMounted.current ) {
+				const newMapOutput = latestMapSelect.current( registry.select, registry );
```

Comment:
So So if I'm reading this properly, the mapping function is never called when the mapping function changes or when the props used in the mapping function changes. This feels like a bug to me, what if a prop changes and the store wasn't updated in the meantime.

I guess what I'm saying here is I'm still not sure why we dropped the dependencies array from the arguments.

Reply:
> So if I'm reading this properly, the mapping function is never called when the mapping function changes 

I think it is:

```js
if ( latestMapSelect.current !== mapSelect ) {
```

`latestMapSelect` gets set to the current mapping function on ever render and the next render compares that value with the previous.  If they differ, then the function is invoked again to get the mapOutput.

> I guess what I'm saying here is I'm still not sure why we dropped the dependencies array from the arguments.

The dependency is now on the `mapSelect` callback.  This gives more control to developers to optimize on their mapSelect function.

I think there's two approaches we could go here:

### `useSelect` _requires_ a dependency array and uses that to optimize rendering.

Disadvantages:
- withSelect` will be harder to optimize for because `ownProps` fluctuates too much so we can't just do `Object.values( ownProps )`
- Easier for developers to shoot themselves in the foot because their dependency array doesn't correctly all the dependencies in their `mapSelect` callback (granted that might be somewhat resolveable via an ES lit rule).

Advantages:
- enforces optimal approaches for performance.
- makes the dependencies more clear and obvious.

### Developers do their own memoization of the `mapSelect` callback.

Disadvantages:

- performance responsibility is shifted to developers, so they can still shoot themselves in the foot wrt performance (granted, something being non performant via not working are too different things :) )

Advantages:

- More flexibility for developers with regards to how they optimize things.
- `withSelect` is a bit more easier to optimize (although there's still some difficulties as noted in the comments for my latest commit).

Reply:
I have a strong preference for having a `dependencies` argument in the signature, mimic `useEffect`, and absorb the optimizations in the framework personally. 

Will we need more fine-tuning? Maybe, but it seems like this could be a `useSelectAdvanced` similar to how there's a `connectAdvanced` in `react-redux`

Reply:
I think the connector's job is to update the component when it has to update. Memoizing a selector for performance is a separate concern.

Keep in mind that a lot of the time, specially for computationally expensive selectors, they will already be defined elsewhere without closures over props, removing the need for inline hook callback memoization. This codebase is a good example of that.

I also like the fact that the public API remains simple and developers have one less thing to learn. They already have to learn how to use primitive memoization hooks, so we can build on that.

Let's look at a few examples:

```js
// Callback never changes.
// Performance is good.
useSelect( importedSelector );

// Callback changes and re-runs on every render.
// `.getItem` is memoized.
// Performance is good.
useSelect( ( select ) => select( 'store' ).getItem() );

// Callback changes and re-runs on every render.
// `.getItem` is memoized with `prop`.
// Performance is good.
useSelect( ( select ) => select( 'store' ).getItem( prop ) );

// Callback changes and re-runs on every render.
// `.getItem` is not memoized.
// Performance suffers.
// We memoize ourselves.
useSelect( useCallback( ( select ) => select( 'store' ).getItem( prop ), [ prop ] ) );

// Callback changes and re-runs on every render.
// `.getItem` is not memoized.
// Performance suffers.
// We delegate memoization to the implementation.
useSelect( ( select ) => select( 'store' ).getItem( prop ), [ prop ] );

```

See where the two approaches differ in the last two calls. **My main issue with having the `dependencies` array is that the user has to learn a new interpretation of it. Does it memoize the callback everywhere? Only for renders? Does it always use the latest one in store updates? Does it update the memoized value after that?**

It very quickly makes it much harder to think about and learn. Developers already know how `useCallback` works and it's much easier to say: This hook calls your callback every time it changes and when the store updates, and then returns its output. If your callback is expensive, you can memoize it, because it's only called when it changes or when the store updates.

Reply:
> I think the connector's job is to update the component when it has to update. Memoizing a selector for performance is a separate concern.

The same thing could have been said about `useEffect` hook. Why the React team decided to implement the dependencies there and not rely on `useCallback` to ensure it runs the effect only when the callback changes? 

I understand that a "pure" implementation is better for an advanced developper that understands how to override the default behavior and build on top of it but let's take a look at the common use-case from a user's perspective:

```
const { block } = useSelect( ( select ) => {
  return {
    block: select( 'core/editor' ).getBlock()
    // call other selectors
  };
} );
```

In our previous attempts to improve the performance of the editor, the fact that the selectors were memoized didn't prove to be enough to address the editor performance issues once we load a lot of blocks. The problem was that a lot of selectors were being run (memoized or not) on each subscribe. So for me, the more we avoid running selectors entirely, the better is.

I understand that I'm basing this on previous metrics, that might not transpose 1-1 to the current use-cases. Ideally, we'd perform performance tests and compare. 

> My main issue with having the dependencies array is that the user has to learn a new interpretation of it

That's my main issue with the dependencies-less implementation as well. Users need to learn how to memoize things by them selves (using other hooks) while the dependencies array is a common pattern already used in `useEffect`.

---

Ultimately, I think we should take a path, mark it as experimental and evaluate it with real usage.

Reply:
In thinking about this more. It seems that the behavior of both approaches is the same if we omit passing any second argument and if we don't check for the callback refrence (always run the callback on render and rely on memoization), that suggests that we could potentially adopt a first version without any attempt to memoize (based on callback or based on dependencies) as a common ground and adapt using concrete measures.

Reply:
The functionality of `useEffect` is completely different. It's not just about closures. There you might want to run the effect more or less times for other reasons. That's why they didn't rely on `useCallback`.

This is really just about:

```js
useSelect( mapSelect, [ dep ] )
```
vs.
```js
useSelect( useCallback( mapSelect, [ dep ] ) )
```

The first one does have less code and I guess it's fine if we document it well enough. Something like:

>This hook calls your callback every time it changes and when the store updates, and then returns its output. If your callback is expensive, you can pass a dependency array as the second parameter to memoize it, because we will only call it when it changes or when the store updates.

I updated the gist: https://gist.github.com/epiqueras/7eae39ba6b903286cf17a4907902a630.